Add buildDubPackage and dub-to-nix for building dub based packages

TomaSajt opened the pull request:
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/301302

This PR adds three main things:
- the `buildDubPackage` helper function
- the `dub-to-nix` helper script
- the `makeDubDep` utility function to be used inside the files generated by `dub-to-nix`

The idea first came up in https://github.com/NixOS/nixpkgs/pull/288841 where I ended up implementing it. I though it would be a better idea to separate those changes into different PR, which is this one.

I also ended up migrating all 2 packages that use `dub` inside of nixpkgs.
`Literate` didn't even have anything to lock with `dub-to-nix`, so it has an empty dependency list.
The changes done to `btdu` is much more visible, though. The package declaration became much shorter, even when it only had 4 packages to lock.

I removed `btdu`'s update script for now, maybe when I implement something like `passthru.fetch-deps` from `buildDotnetModule` for `buildDubPackage` I'll add a new one back.

I inited the `serve-d` package in this PR too to showcase that this works with many dependencies too. 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

TomaSajt committed:
ldc: add meta.mainProgram

Atemu commented on pkgs/build-support/dlang/dub-to-nix/default.nix:
```diff
@@ -0,0 +1,66 @@
+    nix
+  ];
+in
+writeShellScriptBin "dub-to-nix" ''
+  set -euo pipefail
```
Perhaps use writeShellApplication, it sets these by default and does shellcheck.

AFAICT, all this needs passed is the PATH, so this could probably be spun out into a separate file for easier editing.

Atemu replied:
Much cleaner, thanks :)

Atemu commented on pkgs/development/tools/literate-programming/Literate/default.nix:
```diff
@@ -1,28 +1,48 @@
+  '';
 
-  installPhase = "install -D bin/lit $out/bin/lit";
+  doCheck = true;
```
Not sure how the other derivation wrappers do this but if this is this generally desirable for dub packages, perhaps it could be enabled by default.

Atemu commented on pkgs/development/tools/literate-programming/Literate/default.nix:
```diff
@@ -1,28 +1,48 @@
+  doCheck = true;
 
-  meta = with lib; {
+  checkPhase = ''
+    runHook preCheck
+    dub test
+    runHook postCheck
+  '';
```
This would probably make sense to have in buildDubPackage.

TomaSajt replied:
It turns out `Literate` didn't actually have any tests, but had a `Makefile` entry for testing. I'll try to build `serve-d` to test if the checking logic actually works.

Atemu commented on pkgs/development/tools/literate-programming/Literate/default.nix:
```diff
@@ -1,28 +1,48 @@
+    runHook postCheck
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    install -Dm755 bin/lit -t $out/bin
+    runHook preInstall
+  '';
```
Is there no standardised install command for dub?

jtbx replied:
No, but I can file an issue asking about it. I feel like such a feature would be useful to me personally anyway.

TomaSajt replied:
There exist an optional `targetPath` value inside `dub.sdl` or `dub.json`, AFAICT defaulting to `./.`. It's where it puts the final binaries and the files inside `copyPath`.

Perhaps some `jq` based manipulation could be done to the file to add/update it and set it to a fixed directory, though I am not certain if this is robust enough.
(You can convert from `.sdl` to `.json` with a `dub` command, so `jq` could be used in either case)

symphorien replied:
alternatively this could be an argument to buildDubPackage

Atemu reviewed (approved):
This looks like a fine first iteration.

Could you somehow make clear that you (and perhaps @jtbx too) maintain the dlang support? A comment on top of the files or even setting yourself as the codeowner.

Some smaller things and questions:

jtbx commented:
> Could you somehow make clear that you (and perhaps @jtbx too) maintain the dlang support? A comment on top of the files or even setting yourself as the codeowner.

I would be happy to :)

TomaSajt commented:
Added a `CODEOWNERS` entry for D

By the way, what should we call this language internally? I am using `dlang`for `build-support`, because IMO `d` is way too short.

Atemu commented:
Ugh, I just remembered an annoying limitation of GH CODEOWNER: You two must have commit access.

I'd say add a comment in the files instead.

TomaSajt commented:
> Ugh, I just remembered an annoying limitation of GH CODEOWNER: You two must have commit access.
> 
> I'd say add a comment in the files instead.

That's a bit annoying, yes. I'll add some comments, then.

TomaSajt commented:
I added some extra builder options for `makeDubPackage`. Previously it was only `dubFlags`, but now we have `dubFlags` `dubBuildFlags` and `dubTestFlags`.
I also added `combined*Deps` to factor out most of the string interpolation part from the script definition.

I created `pkgs/build-support/dlang/README.md` with a list of maintainers.

TomaSajt commented:
Okay, so I set `doCheck` to `false` by default instead of `true`, because most applications don't have proper testing, and the fallback action when running `dub test` without proper unittests being set up is not very desirable. People will have to opt-in to checking by setting `doCheck = true;`.

AndersonTorres commented on pkgs/top-level/all-packages.nix:
```diff
@@ -7577,6 +7577,9 @@ with pkgs;
   dub = callPackage ../development/tools/build-managers/dub { };
 
+  inherit (callPackage ../build-support/dlang/dub-support.nix { })
+    buildDubPackage dub-to-nix makeDubDep;
```
Why not `dub2nix`?

TomaSajt replied:
The original choice for the name was because this was based off of `nuget-to-nix`, and I forgot to check whether other locker scritps used `-to-` or `2`.
The second reason is that after creating this helper, I found out about the existing `dub2nix` repository, which is an IFD based non-shell application which also does something similar to this. I kept the name as `dub-to-nix` to make the two things distinct.
Although, I am not opposed to changing the name.

symphorien commented on pkgs/build-support/dlang/dub-to-nix/dub-to-nix.sh:
```diff
@@ -0,0 +1,43 @@
+trap 'rm -r "$tmp"' EXIT
+
+if [ ! -f dub.selections.json ]; then
+    echo "The file \`dub.selections.json\` does not exist" >&2
```
```suggestion
    echo "The file \`dub.selections.json\` does not exist in the current working directory" >&2
```

symphorien commented on pkgs/build-support/dlang/dub-to-nix/dub-to-nix.sh:
```diff
@@ -0,0 +1,43 @@
+done
+
+echo "{ makeDubDep }:"
+echo "["
```
current approach is fine but maybe it's less fragile to output a json file with just data (pname, version, sha256), and then move the logic in buildDubPackage to:
-  load the json with builtins.fromJSON and builtins.readFile
- map makeDubDep on it

also consider running shellcheck on this script

TomaSajt replied:
I like this idea. Essentially, it would just be a better `dub.selections.json`. Time to do some more `jq` magic.

symphorien replied:
Instead of doing jq magic you could also write the script in a language that supports json out of the box like python ;)

TomaSajt replied:
I did it in python :)
I'm not sure what's the cleanest way of turning it into an executable, though.

symphorien reviewed (commented):
Thank you for your work!

I'm not very familiar with D, but when compiling d-serve I get these warnings:
```
     Warning Selected package libdparse 0.19.4 does not match the dependency specification >=0.14.0 <0.18.0 in package dfmt. Need to "dub upgrade"?
     Warning Selected package dscanner 0.12.2 does not match the dependency specification 0.11.1 in package workspace-d. Need to "dub upgrade"?
     Warning Selected package dsymbol 0.11.3 does not match the dependency specification 0.13.0 in package dscanner. Need to "dub upgrade"?
     Warning Selected package libdparse 0.19.4 does not match the dependency specification >=0.14.0 <0.18.0 in package dsymbol. Need to "dub upgrade"?
     Warning Selected package libddoc 0.7.4 does not match the dependency specification ~>0.8.0 in package dscanner. Need to "dub upgrade"?
     Warning Selected package libdparse 0.19.4 does not match the dependency specification >=0.13.0 <0.18.0 in package libddoc. Need to "dub upgrade"?
     Warning Selected package libdparse 0.19.4 does not match the dependency specification 0.19.2 in package dscanner. Need to "dub upgrade"?
     Warning Selected package libdparse 0.19.4 does not match the dependency specification 0.17.0 in package workspace-d. Need to "dub upgrade"
```
is it due to dub-to-nix?


Also, a pet peeve of mine: can you set strictDeps to true by default like for python, ocaml, go and a few other language specific builders? this should optimally be true by default all the time, but we keep false by default to avoid breaking everything. In the case of the d ecosystem in nixpkgs, we are starting from nearly zero so let's use the occasion to have good defaults.

TomaSajt commented:
> Thank you for your work!
> 
> I'm not very familiar with D, but when compiling d-serve I get these warnings:
> 
> ```
>      Warning Selected package libdparse 0.19.4 does not match the dependency specification >=0.14.0 <0.18.0 in package dfmt. Need to "dub upgrade"?
>      Warning Selected package dscanner 0.12.2 does not match the dependency specification 0.11.1 in package workspace-d. Need to "dub upgrade"?
>      Warning Selected package dsymbol 0.11.3 does not match the dependency specification 0.13.0 in package dscanner. Need to "dub upgrade"?
>      Warning Selected package libdparse 0.19.4 does not match the dependency specification >=0.14.0 <0.18.0 in package dsymbol. Need to "dub upgrade"?
>      Warning Selected package libddoc 0.7.4 does not match the dependency specification ~>0.8.0 in package dscanner. Need to "dub upgrade"?
>      Warning Selected package libdparse 0.19.4 does not match the dependency specification >=0.13.0 <0.18.0 in package libddoc. Need to "dub upgrade"?
>      Warning Selected package libdparse 0.19.4 does not match the dependency specification 0.19.2 in package dscanner. Need to "dub upgrade"?
>      Warning Selected package libdparse 0.19.4 does not match the dependency specification 0.17.0 in package workspace-d. Need to "dub upgrade"
> ```
> 
> is it due to dub-to-nix?

I believe it's an upstream issue: they have a `dub.selections.json`, but the selected versions are not in the expected version bounds... I have no idea why, though. We could regenerate the selections file, that would get rid of the warnings.

TomaSajt commented:
Okay, I pushed a commit which moves the current system over to a JSON-based lockfile format.

Now there are several questions that might come up:
- do we put the dependencies in non-top level list/set in the lockfile? (`dependencies` in this case)
  - if yes, should we also introduce a `lockfileFormat` versioning number
- should the lockfile be `{pname: { version, sha256 }, ...}` or `[ { pname, version, sha256 } ... ]`
- at which point should the lockfile be read?
  - do we just take in a path to the lockfile and read it internally
    - this is more obvious, easier for the user
  - or do we require user to manually call `lib.importJSON`
    - not that obvious
    - the user could append extra dependencies using nix, which would get rid of the current need for `extraDubDeps` and non-internal usages of `makeDubDep`
      - though something like `extraDubDeps` might be needed in the future if we want to use a non-zip source
- should the file even be JSON?
  - it could be just a nix file instead, but not what it was before (we'd do the mapping internally, like with JSON)
    - not really necessary, would still need to call `import`

symphorien commented:
you forgot to update serve-d to the new lock format.

> do we put the dependencies in non-top level list/set in the lockfile? (dependencies in this case) 

> should the lockfile be {pname: { version, sha256 }, ...} or [ { pname, version, sha256 } ... ]

the nim lock file looks quite alike to your current approach

> or do we require user to manually call lib.importJSON 

I don't see a use case for handling the json value in nix and not directly from the generated file. If you really want to be versatile, you can do like callpackage which accepts a function or a file evaluating to a function. A bit like this:
`let actualData = if builtins.isPath arg then lib.importJSON arg else arg; `

> should the file even be JSON? 

this way you don't need to worry about escaping (what if there is a ${ in a string?). Also such metadata is better off as a pure data format than as a code format. "Old-style" (setup.py) python packaging format being python code has led to inordinate amount of pain and abuse. The ecosystem is moving to a pure data format (toml).

TomaSajt commented:
> > or do we require user to manually call lib.importJSON
> 
> I don't see a use case for handling the json value in nix and not directly from the generated file. If you really want to be versatile, you can do like callpackage which accepts a function or a file evaluating to a function. A bit like this: `let actualData = if builtins.isPath arg then lib.importJSON arg else arg; `

I like this idea, this way I can remove `extraDubDeps`.



> you forgot to update serve-d to the new lock format.

It looks like I accidentally committed it once (I didn't mean to).

TomaSajt committed:
buildDubPackage, dub-to-nix: init

TomaSajt committed:
Literate: use buildDubPackage

TomaSajt committed:
btdu: use buildDubPackage

TomaSajt committed:
serve-d: init at 0.7.6

TomaSajt commented:
Okay I added an actual commit for `serve-d` (but if it's not in the scope of the PR, I'll move it to another PR)

Other than that, I removed `extraDubDeps` and I moved `makeDubDep` to be internal-only. We can expose `makeDubDep` in the future, but it's unlikely that that's ever going to be needed.

I added some comments for the build helper's parameters, even if it's not proper docs, it exists at least.

symphorien commented on pkgs/build-support/dlang/builddubpackage/default.nix:
```diff
@@ -0,0 +1,129 @@
+stdenv.mkDerivation (
+  builtins.removeAttrs args [
+    "dubDeps"
+    "extraDubDeps"
```
I think you forgot to remove extraDubDeps here

TomaSajt replied:
Yeah, I forgot about this completely. Now only `dubLock` is in the list.
However, the other inputs will get put into the builder as environment variables (if `__structuredAttrs` is not enabled). Should those also be filtered out?

Or maybe we could keep them and use `finalAttrs.xyz` to reference them, so that we can do `overrideAttrs` on the package. Not sure about how much builders should support overriding ATM. I've read about a builder helper unification plan, but I think it's not going to happen soon.

symphorien reviewed (commented):
looks fine.

is there a way to tell dub to respect $NIX_BUILD_CORES ?

imo docs and a release notes item are required to merge.

TomaSajt commented:
> is there a way to tell dub to respect $NIX_BUILD_CORES ?

there is `--build-mode=singleFile --parallel` which will separate the workload onto multiple cores, however we can't specify the exact number of cores.

TomaSajt commented:
Okay, I updated https://github.com/NixOS/nixpkgs/pull/288841 to the new lockfile format, here's how you could do extra dub dependencies:
https://github.com/NixOS/nixpkgs/blob/dae2ee5c84d647add18220d5dbf14a4d8784cfd3/pkgs/applications/misc/inochi2d/generic.nix#L62-L73

TomaSajt committed:
doc: add dlang language section

TomaSajt committed:
nixos/doc: add release notes for buildDubPackage

TomaSajt commented:
Okay, I added both release notes and docs for the things added by this PR (mostly based off of the `nim` section)

symphorien commented on doc/languages-frameworks/dlang.section.md:
```diff
@@ -0,0 +1,65 @@
+    hash = "sha256-3sSZq+5UJH02IO0Y1yL3BLHDb4lk8k6awb5ZysBQciE=";
+  };
+
+  dubLock = ./dub-lock.json;
```
```suggestion
  # generated by dub-to-nix, see below
  dubLock = ./dub-lock.json;
```

symphorien commented on doc/languages-frameworks/dlang.section.md:
```diff
@@ -0,0 +1,65 @@
+
+The following parameters are specific to `buildDubPackage`:
+
+* `dubLock`: The `dub` lockfile. Can be either a path to the file, or an attrset already parsed with `lib.importJSON`.
```
```suggestion
* `dubLock`: A lockfile generated by `dub-to-nix` from the source of the package. Can be either a path to the file, or an attrset already parsed with `lib.importJSON`.
```

symphorien commented on doc/languages-frameworks/dlang.section.md:
```diff
@@ -0,0 +1,65 @@
+* `dubTestFlags ? []`: The flags to pass to `dub test`.
+* `compiler ? ldc`: The D compiler to be used by `dub`.
+
+## Lockfiles {#dub-lockfiles}
```
I think this section should go before the example. There are two steps: 1/generate the lock file, 2/ write nix code like in this example

TomaSajt replied:
I just swapped the lockfile and the parameters part around.
I think new the comment above `dubLock` in the example is enough indication of what to read next.

symphorien commented on doc/languages-frameworks/dlang.section.md:
```diff
@@ -0,0 +1,65 @@
+The following parameters are specific to `buildDubPackage`:
+
+* `dubLock`: The `dub` lockfile. Can be either a path to the file, or an attrset already parsed with `lib.importJSON`.
+  This is useful if you want to use `dub` dependencies not already in the lockfile.
```
I think an example and an explanation of why this could happen could be useful.

TomaSajt replied:
Added an example

symphorien commented on doc/languages-frameworks/dlang.section.md:
```diff
@@ -0,0 +1,65 @@
+## Lockfiles {#dub-lockfiles}
+Nixpkgs has its own lockfile format for `dub`, because `dub`'s official "lockfiles" are not hash based.
+The lockfile is generated using the `dub-to-nix` helper package.
+Running `dub-to-nix` in the root of the source of the program you want to package will print the lockfile to stdout.
```
maybe mention that dub-to-nix is in nixpkgs

```suggestion
Running `nix-shell -p dub-to-nix --run dub-to-nix` in the root of the source of the program you want to package will print the lockfile to stdout.
```

TomaSajt replied:
Did it a bit differently

symphorien reviewed (commented):
thank you for the documentation! I left a few comments

Atemu commented on pkgs/development/tools/literate-programming/Literate/dub-lock.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "dependencies": {}
+}
```
This could be inlined or even just be the default of buildDubPackage.

TomaSajt replied:
It *could* be inlined, but I don't think many packages have 0 dependencies. If we have a default, it won't show a missing attribute error, which is worse IMO.

Atemu reviewed (approved):
Result of `nixpkgs-review pr 299618` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>Literate</li>
    <li>btdu</li>
    <li>dub-to-nix</li>
    <li>serve-d</li>
    <li>tests.haskell.cabalSdist.assumptionLocalHasDirectReference</li>
    <li>tests.haskell.cabalSdist.localHasNoDirectReference</li>
  </ul>
</details>

Successfully ran `btdu`.

Great work! I especially like the docs :)

TomaSajt commented:
Thanks for all the help and reviews, everyone!
This has probably been my most complex contribution to date, glad to see it get merged.

I'll get back cleaning up my other PR now.
